Reopen "get k8s config from requestConfig. and added test."

hnarimiya opened the pull request:
I rebased this PR(#1279) from master.
Please review.

yoyama commented:
IIUC, this PR will allow to configure K8 in workflow definition ?
If so, I would like to ask you to add server configuration to enable/disable of it and default is disable for server administrators.

yoyama reviewed (changes requested):
Please check my comment

hnarimiya commented:
@yoyama 
Sorry for late reply.

I'm sorry if it's not what I intended.

I would like to add a parameter that allows setting using the parameter of "kubernetes" from the workflow.
The parameters are as follows.
```
agent.command_executor.kubernetes.allow_configure_workflow_definition=true
```
If this parameter is true, requestConfig is merged with systemConfig and used like PR.

If false, only systemConfig is used. (As an image, the original createFromSystemConfig method)

If this parameter is not set or defaults, it is treated as false.

yoyama commented on digdag-standards/src/main/java/io/digdag/standards/command/kubernetes/KubernetesClientConfig.java:
```diff
@@ -19,42 +19,50 @@ public static KubernetesClientConfig create(final Optional<String> name,
+        if (extractedSystemConfig != null && extractedRequestConfig != null) {
+            config = extractedSystemConfig.merge(extractedRequestConfig);
+
+        } else if (extractedRequestConfig != null) {
```
May I ask you to change the format ?
```
}
else if (
```

yoyama commented on digdag-standards/src/main/java/io/digdag/standards/command/kubernetes/KubernetesClientConfig.java:
```diff
@@ -19,42 +19,50 @@ public static KubernetesClientConfig create(final Optional<String> name,
+
+        // Create a config that merges RequestConfig with SystemConfig
+        final Config config;
+        if (extractedSystemConfig != null && extractedRequestConfig != null) {
```
If you set empty Config as default value in both `extractedSystemConfig` and `extractedRequestConfig`.
You can simply merge them without `if ... else if ...` ?
And at last check required keys are set.

yoyama commented on digdag-standards/src/main/java/io/digdag/standards/command/kubernetes/KubernetesClientConfig.java:
```diff
@@ -19,42 +19,50 @@ public static KubernetesClientConfig create(final Optional<String> name,
+            config = extractedSystemConfig;
+
+        } else {
+            throw new ConfigException("systemConfig and requestConfig does not exist");
```
It might be better to improve the message to make users understand what is wrong.

hnarimiya commented:
@yoyama 
Added parameter.
https://github.com/treasure-data/digdag/pull/1692#issuecomment-1175773307

I don't know if this is intended by yoyama san, so please tell me your opinion.

I fixed it based on some indications in the code.
https://github.com/treasure-data/digdag/pull/1692#discussion_r914495928
https://github.com/treasure-data/digdag/pull/1692#discussion_r914499547
https://github.com/treasure-data/digdag/pull/1692#discussion_r914501498

Please review.

yoyama commented:
@hnarimiya Could you rebase your PR to the laster master and fix the CI failures ?

hnarimiya commented:
@yoyama @szyn 
I'm sorry for being late.
Please review.
